pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,662 @@
+		Ok(now >= maturity_date.ensure_add(overdue_secs)?)
+	}
+
+	pub fn find_best(
+		policy: impl Iterator<Item = WriteOffState<Rate>>,
+		maturity_date: Moment,
+		now: Moment,
+	) -> Option<WriteOffState<Rate>> {
+		policy
+			.filter_map(|p| p.is_not_overdue(maturity_date, now).ok()?.then_some(p))
+			.max_by(|a, b| a.overdue_days.cmp(&b.overdue_days))
+	}
+
+	pub fn status(&self) -> WriteOffStatus<Rate> {
+		WriteOffStatus {
+			percentage: self.percentage,
+			penalty: self.penalty,
+		}
+	}
+}
+
+/// Diferent kinds of write off status that a loan can be
+#[derive(Encode, Decode, Clone, PartialEq, Default, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct WriteOffStatus<Rate> {
+	/// Percentage of present value we are going to write off on a loan
+	pub percentage: Rate,
+
+	/// Additional interest that accrues on the written down loan as penalty per sec
+	pub penalty: Rate,
+}
+
+impl<Rate> WriteOffStatus<Rate>
+where
+	Rate: FixedPointNumber,
+{
+	pub fn write_down<Balance: tokens::Balance + FixedPointOperand>(
+		&self,
+		debt: Balance,
+	) -> Result<Balance, ArithmeticError> {
+		debt.ensure_sub(self.percentage.ensure_mul_int(debt)?)
+	}
+
+	pub fn is_none(&self) -> bool {
+		self.percentage.is_zero() && self.penalty.is_zero()
+	}
+}
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time
+	Fixed(Moment),
+}
+
+impl Maturity {
+	pub fn date(&self) -> Moment {
+		match self {
+			Maturity::Fixed(moment) => *moment,
+		}
+	}
+}
+
+/// Interest payment periods
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+}
+
+/// Specify the paydown schedules of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PayDownSchedule {
+	/// The entire borrowed amount is expected to be paid back at the maturity date
+	None,
+}
+
+/// Specify the repayment schedule of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Moment) -> bool {
+		self.maturity.date() > now
+	}
+}
+
+/// TODO
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct DiscountedCashFlows<Rate> {
+	/// TODO
+	probability_of_default: Rate,
+
+	/// TODO
+	loss_given_default: Rate,
+
+	/// TODO
+	discount_rate: Rate,
+}
+
+impl<Rate: FixedPointNumber> DiscountedCashFlows<Rate> {
```

Comment:
I think organizationally, it makes sense to separate the valuation logic from the rest of the types. This will get a lot more complex as we implement repayment schedules with interest payments and non bullet pay down schedules.

Reply:
Agree! I started with all types in one place, but the idea was to split them later 👍🏻 .

Not sure if split also types that know about `T: Config` and types than not. WDYT?

Regarding these parts, there are some TODOs in the comments that I don't sure how to fill with the correct jargon.